gc.c

```diff
@@ -5210,15 +5210,22 @@ static void invalidate_moved_page(rb_objspace_t *objspace, struct heap_page *pag
 
 #if GC_CAN_COMPILE_COMPACTION
 static void
-read_barrier_handler(uintptr_t address)
+read_barrier_handler(uintptr_t original_address)
 {
     VALUE obj;
     rb_objspace_t * objspace = &rb_objspace;
 
-    address -= address % BASE_SLOT_SIZE;
+    /* Calculate address aligned to slots. */
+    uintptr_t address = original_address - (original_address % BASE_SLOT_SIZE);
 
     obj = (VALUE)address;
 
+    /* If the address is 0, then the read barrier cannot handle it and will
+     * segfault. */
+    if (address == 0) {
+        rb_bug("read_barrier_handler: segmentation fault at 0x%lx", original_address);
+    }
```

Comment:
I think we should check the address that's passed to `mprotect` rather than this address.  I think it's possible for `GET_PAGE_BODY` to return 0 even if `address` is not 0, right?